### ISSUE TITLE
Suggesting new "hot-fix" strategy

### ISSUE THREAD

**Issue by htgoebel** [pyinstaller/pyinstaller#1756]
Hi, I'm suggesting a new release strategy for bug-fixes:

**Four (w) weeks after the release, we will publish a bug-fix release, if there are any bug-fixes.**

_Process_
- After a release, we prepare a branch like "hotfix/3.1". All pure bug-fixes should go into this branch. If a change does not only contain a bug-fix, it should be split up.
- Every change on the hot-fix branch will be merged to `develop` immediately.
- The bug-fix will be released 4 weeks after the main release. This is done like any normal release, except it starts at the hotfix-branch.

_Rational_
- Currently we have no strategy for how to release bug-fixes.
- After release 3.0 it showed up that we need to provide bug-fixes much quicker than we did. There have been several severe bugs crashing PyInstaller 3.0, but we solved them on `develop` only. In consequence some projects required building with the development version (e.g. borgbackup, see [this comment](https://github.com/borgbackup/borg/blob/458fc747a3171287ec1091601db6c679d8b1075e/Vagrantfile#L214) - which is not a good thing.
- This process should allow publishing a bug-fix release much easier, since only a few changes should be here and they will (hopefully) not change any.

What do you thing?

**Comment by nchammas** [comment]
I've seen extremely active projects like [Apache Spark](https://github.com/apache/spark) and [Rust](https://github.com/rust-lang/rust) use a ["train" release model](http://blog.rust-lang.org/2014/10/30/Stability.html).

Every N weeks/months they make a stable release based on what is in `master`. Generally, releases are not blocked on new features still in the works. This means that users get updates at a regular pace.

If there are critical bug fixes that need to get out, they make bug fix releases as necessary, without a fixed schedule.

To provide an example, say PyInstaller moves to a 4-month release cycle:
* PyInstaller 3.2 comes out roughly in May, 3.3 comes out in September, and so forth. These releases are generally not blocked on any specific features; rather, they are released at roughly regular intervals.
* If any critical bug fixes need to get out, they are released as necessary. That means 3.2.1, 3.2.2, etc. could come out at any time they are needed while we are still in the development window for 3.3.

**Comment by codewarrior0** [comment]
Yes, we definitely need to do bug-fix releases. I agree that after every stable release, we make a new branch to apply bugfixes to that release. We should put out the bugfix releases as they are needed instead of on a fixed schedule to avoid having a broken version on PyPI for too long.

For example, [Cython](https://github.com/cython/cython/branches) does all the main development on the `master` branch, and for each major release another branch is created (`0.23.x` etc) to apply bugfixes onto. The bugfix releases are also [tagged](https://github.com/cython/cython/releases) to keep track of them more easily.

> Every change on the hot-fix branch will be merged to develop immediately.

Alternately, if the bugfix is committed to `develop` first, it can simply be cherry-picked onto the hot-fix branch.

**Comment by ThomasWaldmann** [comment]
+1

**Comment by matysek** [comment]
* @htgoebel  I second doing bugfix releases the way you suggest.
* If we see any any bug report that is only bugfix short after release, then put that into `3.1.1` milestone and backport it into `develop`.

Lately I have been thinking about doing release every 3 months. I think as long as our test suite passes then it should be ok do do release at any time. I suggest:
* stick to do a release every 3-4 months (3-4 releases per year)

**Comment by htgoebel** [comment]
To summarize your statements, as far as I understood:
- Fixed release intervals (3 months) for normal releases (this is out of scope of this issue, though :-)
- Bug-fix releases when needed
- cherry-picked from `develop`.

So the new proposal is:

**We will publish a bug-fix release, if there are any urgent bugs to fix.**

*Process*
* After a release, we prepare a branch like "release/3.1.1". All pure bug-fixes should go into this branch. If a change does not only contain a bug-fix, it should be split up.
* Fixes can be applied to the `develop` branch as usual and will be cherry-picked to the hot-fix branch as needed.
* The bug-fix release is done like any normal release, except it starts at the hotfix-branch. (Details to be defined when we have some experiences.)
* If no bug-fix release has been made until the next regular release, the resp. branch is simply deleted (even if some changes are already applied to it.)

**Comment by Legorooj** [comment]
Closing due to inactivity and hotfixes being implemented (as far as I can tell)

**Comment by ThomasWaldmann** [comment]
https://github.com/pyinstaller/pyinstaller/releases as one can see there, there are usually no hotfix releases, two .1 patch releases in 2017 being rather the exception than the rule.

So I do not consider this solved.

**Comment by Legorooj** [comment]
Good point. I'll reopen if @htgoebel agrees. This either needs attention, or no attention - ie forget the idea. It's a good idea though.

**Comment by ThomasWaldmann** [comment]
A good way to deal with this would be:
- after x.y(.0) release, create a x.y-maint branch
- patch releases, like x.y.1, x.y.2, etc. are tags within that x.y-maint branch
- if something serious is fixed, it needs backporting to that x.y-maint branch (except if it does not apply)

**Comment by Legorooj** [comment]
Maybe. I'll reopen - it needs discussion.

**Comment by bjones1** [comment]
Doing this requires work by the maintainers. If funded, hotfixes might be possible.

**Comment by Legorooj** [comment]
@bjones1 (groan) and we come straight back round to the funding issue.
